consensus/state.go

```diff
@@ -1644,19 +1644,38 @@ func (cs *ConsensusState) addVote(vote *types.Vote, peerID p2p.ID) (added bool,
 func (cs *ConsensusState) signVote(type_ byte, hash []byte, header types.PartSetHeader) (*types.Vote, error) {
 	addr := cs.privValidator.GetAddress()
 	valIndex, _ := cs.Validators.GetByAddress(addr)
+	timestamp := cs.voteTime()
+
 	vote := &types.Vote{
 		ValidatorAddress: addr,
 		ValidatorIndex:   valIndex,
 		Height:           cs.Height,
 		Round:            cs.Round,
-		Timestamp:        time.Now().UTC(),
+		Timestamp:        timestamp,
 		Type:             type_,
 		BlockID:          types.BlockID{hash, header},
 	}
 	err := cs.privValidator.SignVote(cs.state.ChainID, vote)
 	return vote, err
 }
 
+func (cs *ConsensusState) voteTime() time.Time {
+	now := types.Now()
+	minVoteTime := now
+	// TODO: We should remove next line in case we don't vote for v in case cs.ProposalBlock == nil,
+	// even if cs.LockedBlock != nil. See https://github.com/tendermint/spec.
```

Comment:
In the spec we don't vote for lockedBlock in case we don't agree with Proposal. In case we decide to align with the spec (which I would suggest as Prevoting always for lockedBlock is optimisation which is not clear how much it brings), then the logic here can be simplified as it will only depend on ProposalBlock.